### PR TITLE
Simplify expr.fieldRead

### PR DESCRIPTION
**Pull request by nwt**

### CODE REVIEW

File: expr/fieldexpr.go [hunk 1]
```diff
@@ -42,39 +42,19 @@ func (ai *arrayIndex) apply(e zng.Value) zng.Value {
...
-				}
-				var err error
-				v, _, err = it.Next()
-				if err != nil {
-					return zng.Value{}
-				}
-			}
-			return zng.Value{col.Type, v}
-		}
+	fv, err := zng.NewRecord(recType, v.Bytes).ValueByField(string(f))
```
Comment (mccanne) [inline expr/fieldexpr.go:53]: This creates more GC memory traffic in the inner loop...

Comment (nwt) [reply expr/fieldexpr.go:53]: It doesn't do that (`fv` doesn't escape), but it is slower. I'll speed it up or close this.

* Performance of master at e6dc68ead8bed5e9ae436b7b2cf62e4e9bdec923
```
$ for i in 1 2 3 4 5; do /usr/bin/time ./dist/zq -t 'id.orig_p = 1000' zq-sample-data/zng/conn.zng.gz; done
        1.42 real         1.46 user         0.04 sys
        1.26 real         1.45 user         0.04 sys
        1.25 real         1.44 user         0.04 sys
        1.25 real         1.45 user         0.04 sys
        1.24 real         1.44 user         0.04 sys
```

* Performance of this branch at e6dc68ead8bed5e9ae436b7b2cf62e4e9bdec923
```
$ for i in 1 2 3 4 5; do /usr/bin/time ./dist/zq -t 'id.orig_p = 1000' zq-sample-data/zng/conn.zng.gz; done
        1.60 real         1.47 user         0.04 sys
        1.29 real         1.48 user         0.04 sys
        1.29 real         1.49 user         0.05 sys
        1.27 real         1.47 user         0.04 sys
        1.27 real         1.47 user         0.04 sys
```

* expr/fieldexpr.go optimization details from `go build -gcflags=-m expr/fieldexpr.go` (relevant line: `expr/fieldexpr.go:53:26: &zng.Record literal does not escape`)
```
# command-line-arguments
expr/fieldexpr.go:53:26: inlining call to zng.NewRecord
expr/fieldexpr.go:98:26: inlining call to errors.New
expr/fieldexpr.go:33:7: ai does not escape
expr/fieldexpr.go:33:29: leaking param: e to result ~r1 level=0
expr/fieldexpr.go:47:7: f does not escape
expr/fieldexpr.go:47:26: leaking param: v to result ~r1 level=0
expr/fieldexpr.go:53:26: &zng.Record literal does not escape
expr/fieldexpr.go:69:23: leaking param content: node
expr/fieldexpr.go:104:14: leaking param content: r
expr/fieldexpr.go:70:6: moved to heap: ops
expr/fieldexpr.go:71:6: moved to heap: field
expr/fieldexpr.go:70:26: make([]fieldop, 0) escapes to heap
expr/fieldexpr.go:89:27: []fieldop literal escapes to heap
expr/fieldexpr.go:89:28: &arrayIndex literal escapes to heap
expr/fieldexpr.go:92:27: []fieldop literal escapes to heap
expr/fieldexpr.go:92:37: fieldRead(op.Param) escapes to heap
expr/fieldexpr.go:95:27: ... argument does not escape
expr/fieldexpr.go:95:55: op.Fn escapes to heap
expr/fieldexpr.go:98:26: &errors.errorString literal escapes to heap
expr/fieldexpr.go:104:9: func literal escapes to heap
expr/fieldexpr.go:122:24: leaking param content: nodes
expr/fieldexpr.go:125:19: make([]FieldExprResolver, 0, len(nodes)) escapes to heap
expr/fieldexpr.go:138:24: leaking param content: node
expr/fieldexpr.go:145:22: ... argument does not escape
expr/fieldexpr.go:145:51: FieldExprToString(node.Field) escapes to heap
expr/fieldexpr.go:147:22: ... argument does not escape
expr/fieldexpr.go:147:50: FieldExprToString(node.Field) escapes to heap
expr/fieldexpr.go:147:68: node.Param escapes to heap
expr/fieldexpr.go:149:22: ... argument does not escape
expr/fieldexpr.go:149:49: FieldExprToString(node.Field) escapes to heap
expr/fieldexpr.go:149:67: node.Param escapes to heap
<autogenerated>:1: leaking param: .this
expr/fieldexpr.go:26:8: leaking param: .anon0
<autogenerated>:1: leaking param: .this
<autogenerated>:1: .this does not escape
expr/fieldexpr.go:47:26: leaking param: v to result ~r1 level=0
```

* expr.fieldRead.apply assembly from `go tool objdump -S -s expr.fieldRead.apply dist/zq`
```asm
TEXT github.com/brimsec/zq/expr.fieldRead.apply(SB) /Users/noah/looky/zq/expr/fieldexpr.go
func (f fieldRead) apply(v zng.Value) zng.Value {
  0x11b6ad0		65488b0c2530000000		MOVQ GS:0x30, CX	
  0x11b6ad9		488d4424c8			LEAQ -0x38(SP), AX	
  0x11b6ade		483b4110			CMPQ 0x10(CX), AX	
  0x11b6ae2		0f86e4010000			JBE 0x11b6ccc		
  0x11b6ae8		4881ecb8000000			SUBQ $0xb8, SP		
  0x11b6aef		4889ac24b0000000		MOVQ BP, 0xb0(SP)	
  0x11b6af7		488dac24b0000000		LEAQ 0xb0(SP), BP	
  0x11b6aff		48c78424f800000000000000	MOVQ $0x0, 0xf8(SP)	
  0x11b6b0b		0f57c0				XORPS X0, X0		
  0x11b6b0e		0f11842400010000		MOVUPS X0, 0x100(SP)	
  0x11b6b16		0f11842410010000		MOVUPS X0, 0x110(SP)	
	recType, ok := v.Type.(*zng.TypeRecord)
  0x11b6b1e		488b8424d8000000	MOVQ 0xd8(SP), AX										
  0x11b6b26		488b8c24d0000000	MOVQ 0xd0(SP), CX										
  0x11b6b2e		488d150b788900		LEAQ go.itab.*github.com/brimsec/zq/zng.TypeRecord,github.com/brimsec/zq/zng.Type(SB), DX	
  0x11b6b35		4839d1			CMPQ DX, CX											
  0x11b6b38		0f8584010000		JNE 0x11b6cc2											
	if !ok {
  0x11b6b3e		0f8541010000		JNE 0x11b6c85		
	fv, err := zng.NewRecord(recType, v.Bytes).ValueByField(string(f))
  0x11b6b44		488b8c24e8000000	MOVQ 0xe8(SP), CX	
  0x11b6b4c		488b9424e0000000	MOVQ 0xe0(SP), DX	
  0x11b6b54		488b9c24f0000000	MOVQ 0xf0(SP), BX	
	return &Record{
  0x11b6b5c		48c744247800000000	MOVQ $0x0, 0x78(SP)	
  0x11b6b65		0f11842480000000	MOVUPS X0, 0x80(SP)	
  0x11b6b6d		0f11842490000000	MOVUPS X0, 0x90(SP)	
  0x11b6b75		0f118424a0000000	MOVUPS X0, 0xa0(SP)	
  0x11b6b7d		48c744247800000000	MOVQ $0x0, 0x78(SP)	
  0x11b6b86		0f11842480000000	MOVUPS X0, 0x80(SP)	
  0x11b6b8e		0f11842490000000	MOVUPS X0, 0x90(SP)	
  0x11b6b96		0f118424a0000000	MOVUPS X0, 0xa0(SP)	
  0x11b6b9e		4889442478		MOVQ AX, 0x78(SP)	
  0x11b6ba3		c684248000000001	MOVB $0x1, 0x80(SP)	
  0x11b6bab		4889942488000000	MOVQ DX, 0x88(SP)	
  0x11b6bb3		48898c2490000000	MOVQ CX, 0x90(SP)	
  0x11b6bbb		48899c2498000000	MOVQ BX, 0x98(SP)	
	fv, err := zng.NewRecord(recType, v.Bytes).ValueByField(string(f))
  0x11b6bc3		488d442478		LEAQ 0x78(SP), AX						
  0x11b6bc8		48890424		MOVQ AX, 0(SP)							
  0x11b6bcc		488b8424c0000000	MOVQ 0xc0(SP), AX						
  0x11b6bd4		4889442408		MOVQ AX, 0x8(SP)						
  0x11b6bd9		488b8424c8000000	MOVQ 0xc8(SP), AX						
  0x11b6be1		4889442410		MOVQ AX, 0x10(SP)						
  0x11b6be6		e8d505ffff		CALL github.com/brimsec/zq/zng.(*Record).ValueByField(SB)	
  0x11b6beb		48837c244000		CMPQ $0x0, 0x40(SP)						
  0x11b6bf1		488b442418		MOVQ 0x18(SP), AX						
  0x11b6bf6		4889442450		MOVQ AX, 0x50(SP)						
  0x11b6bfb		0f10442420		MOVUPS 0x20(SP), X0						
  0x11b6c00		0f11442458		MOVUPS X0, 0x58(SP)						
  0x11b6c05		0f10442430		MOVUPS 0x30(SP), X0						
  0x11b6c0a		0f11442468		MOVUPS X0, 0x68(SP)						
	if err != nil {
  0x11b6c0f		743d			JE 0x11b6c4e		
		return zng.Value{}
  0x11b6c11		488b0568898800		MOVQ github.com/axiomhq/hyperloglog.(*Sketch).mergeSparse.stkobj+320(SB), AX	
  0x11b6c18		48898424f8000000	MOVQ AX, 0xf8(SP)								
  0x11b6c20		0f100561898800		MOVUPS github.com/axiomhq/hyperloglog.(*Sketch).mergeSparse.stkobj+328(SB), X0	
  0x11b6c27		0f11842400010000	MOVUPS X0, 0x100(SP)								
  0x11b6c2f		0f100562898800		MOVUPS github.com/axiomhq/hyperloglog.(*Sketch).mergeSparse.stkobj+344(SB), X0	
  0x11b6c36		0f11842410010000	MOVUPS X0, 0x110(SP)								
  0x11b6c3e		488bac24b0000000	MOVQ 0xb0(SP), BP								
  0x11b6c46		4881c4b8000000		ADDQ $0xb8, SP									
  0x11b6c4d		c3			RET										
	return fv
  0x11b6c4e		488b442450		MOVQ 0x50(SP), AX	
  0x11b6c53		48898424f8000000	MOVQ AX, 0xf8(SP)	
  0x11b6c5b		0f10442458		MOVUPS 0x58(SP), X0	
  0x11b6c60		0f11842400010000	MOVUPS X0, 0x100(SP)	
  0x11b6c68		0f10442468		MOVUPS 0x68(SP), X0	
  0x11b6c6d		0f11842410010000	MOVUPS X0, 0x110(SP)	
  0x11b6c75		488bac24b0000000	MOVQ 0xb0(SP), BP	
  0x11b6c7d		4881c4b8000000		ADDQ $0xb8, SP		
  0x11b6c84		c3			RET			
		return zng.Value{}
  0x11b6c85		488b0534888800		MOVQ github.com/axiomhq/hyperloglog.(*Sketch).mergeSparse.stkobj+128(SB), AX	
  0x11b6c8c		48898424f8000000	MOVQ AX, 0xf8(SP)								
  0x11b6c94		0f10052d888800		MOVUPS github.com/axiomhq/hyperloglog.(*Sketch).mergeSparse.stkobj+136(SB), X0	
  0x11b6c9b		0f11842400010000	MOVUPS X0, 0x100(SP)								
  0x11b6ca3		0f10052e888800		MOVUPS github.com/axiomhq/hyperloglog.(*Sketch).mergeSparse.stkobj+152(SB), X0	
  0x11b6caa		0f11842410010000	MOVUPS X0, 0x110(SP)								
  0x11b6cb2		488bac24b0000000	MOVQ 0xb0(SP), BP								
  0x11b6cba		4881c4b8000000		ADDQ $0xb8, SP									
  0x11b6cc1		c3			RET										
  0x11b6cc2		b800000000		MOVL $0x0, AX									
	recType, ok := v.Type.(*zng.TypeRecord)
  0x11b6cc7		e972feffff		JMP 0x11b6b3e		
func (f fieldRead) apply(v zng.Value) zng.Value {
  0x11b6ccc		e85fc6eaff		CALL runtime.morestack_noctxt(SB)			
  0x11b6cd1		e9fafdffff		JMP github.com/brimsec/zq/expr.fieldRead.apply(SB)	

  0x11b6cd6		cc			INT $0x3		
  0x11b6cd7		cc			INT $0x3		
  0x11b6cd8		cc			INT $0x3		
  0x11b6cd9		cc			INT $0x3		
  0x11b6cda		cc			INT $0x3		
  0x11b6cdb		cc			INT $0x3		
  0x11b6cdc		cc			INT $0x3		
  0x11b6cdd		cc			INT $0x3		
  0x11b6cde		cc			INT $0x3		
  0x11b6cdf		cc			INT $0x3		
```

**Review by mccanne — APPROVED** [review]

**Comment by mccanne** [comment]
@nwt This all changed in branch eval but this is still here... 

https://github.com/brimsec/zq/blob/c807138979746bab234962d33310fa2b40107ac1/expr/fieldexpr.go#L36

should we make the change?

**Comment by mccanne** [comment]
It seems like getNthFromContainer() (with a better name) should move to zng/recordval.go and expr, rec.Slice(), and array acces should all use it.

**Pull request closed without merging**